### PR TITLE
add read_only_columns to TableConfig

### PR DESCRIPTION
**Pull request by sinisaos**
Related to #138 and partially to #271

### CODE REVIEW

**Comment by dantownsend** [comment]
@sinisaos Thanks for this.

It looks like we're hiding inputs if they're marked as readonly, but another option is to set the `readonly` attribute, so they are still visible, just not editable:

https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/readonly

You should be able to do this using Vue:

```
:readonly="isReadOnlyString(columnName)"
```

**Comment by sinisaos** [comment]
> You should be able to do this using Vue:
> 
> ```
> :readonly="isReadOnlyString(columnName)"
> ```

@dantownsend I already tried that but that doesn't work. I think hidding field (to disable the editing of certain form fields) is not a bad solution because we have, in addition to input fields, various selects, DurationWidget, ChoiceWidget, FlatPickr, etc. where the `readonly` attribute does not work and we have to use the `disabled` attribute and we would have to change a lot more code.

**Comment by sinisaos** [comment]
@dantownsend With the latest commit the field should be visible but `read-only` or `disabled` in the edit form. I hope this is better and I hope I have included all the possible fields.

**Comment by dantownsend** [comment]
@sinisaos It's OK - I can run the tests locally. 

You can download the artefacts from GitHub actions which might provide some clues:

https://github.com/piccolo-orm/piccolo_admin/actions/runs/4321016381

**Comment by sinisaos** [comment]
@dantownsend Once again, the Playwright proved to be very useful. I changed a lot of code to pass the test with custom forms because there was a conflict in the schemas. I tried using your approach. Can you check it when you have time?

**Comment by dantownsend** [comment]
@sinisaos Yeah, the Playwright tests are great - I'd like to write more. I will add some for nullable columns soon.

**Comment by sinisaos** [comment]
@dantownsend You don't even have to write tests, you can generate tests with [playwright codegen](https://playwright.dev/python/docs/codegen) and then copy-paste the generated code.

**Comment by dantownsend** [comment]
@sinisaos I don't mind the auto generated tests. When there are lots of tests I like to follow this approach:

https://playwright.dev/docs/pom

File: piccolo_admin/endpoints.py [hunk 1]
```diff
@@ -774,7 +790,7 @@ async def store_file(
 
         try:
             file_key = await media_storage.store_file(
-                file_name=file.filename,
+                file_name=file.filename,  # type: ignore
```
Comment (dantownsend) [inline piccolo_admin/endpoints.py:793]: What's the error with this?

**Comment by dantownsend** [comment]
This looks great. I just need to test it locally.

File: piccolo_admin/example.py [hunk 2]
```diff
@@ -322,6 +322,10 @@ async def booking_endpoint(request, data):
         Movie.release_date,
     ],
     rich_text_columns=[Movie.description],
+    read_only_columns=[
+        Movie.release_date,
+        Movie.won_oscar,
+    ],
```
Comment (dantownsend) [inline piccolo_admin/example.py:328]: One thing we might want to do is create a new table just for testing read only columns. I did something similar for nullable columns:

https://github.com/piccolo-orm/piccolo_admin/blob/2beacfe3d0285fcf0e09a1141ad26121906a91f5/piccolo_admin/example.py#L213-L222

So we have a table with every possible column type, so we can quickly see that they all work. It'll also make it easier when writing Playwright tests.

File: piccolo_admin/endpoints.py (hunk 1 above)
Comment (sinisaos) [reply piccolo_admin/endpoints.py:793]: There was a mypy error for this, but now I tried to clone the branch into a fresh virtualenv, ran linter and the error went away. Sorry, it's my fault. I will remove it.

Comment (dantownsend) [reply piccolo_admin/endpoints.py:793]: No problem - just thought it might be something I needed to fix.

File: piccolo_admin/example.py (hunk 2 above)
Comment (sinisaos) [reply piccolo_admin/example.py:328]: That's a great idea. I will try to make that table with all possible column types and test it.

**Comment by dantownsend** [comment]
Ideally this would be verified on the API level too, but could maybe implement that later on in `PiccoloCRUD`. For now, we might need to add something to the docs warning people that it's the case.

File: piccolo_admin/example.py (hunk 2 above)
Comment (dantownsend) [reply piccolo_admin/example.py]: Thanks

**Comment by dantownsend** [comment]
I would put the `ReadOnlyColumns` table also under the `Testing` group.

Screenshot 2023-03-06 at 21 26 39

**Comment by dantownsend** [comment]
I've tested it. I noticed a couple of things.

Array columns still let you add or remove items:

Screenshot 2023-03-06 at 21 30 15

Read only fields only work on the edit page. So this is a tricky one ... should they also be read only when adding a new row? This means the value will be the default. I think it's useful for columns like the primary key, created on / modified on etc.

**Comment by dantownsend** [comment]
Another suggestion - we could put `read only` next to the label, so it's obvious that it's read only. Otherwise users might think it's just broken.

Screenshot 2023-03-06 at 21 40 16

You can add this CSS to show a special cursor too:

```css
input:read-only {
    cursor: not-allowed !important;
}
```

**Comment by sinisaos** [comment]
@dantownsend I added a read-only to field label and disabled links for `ArrayWidget` in read-only mode.

> Read only fields only work on the edit page. So this is a tricky one ... should they also be read only when adding a new row? This means the value will be the default. I think it's useful for columns like the primary key, created on / modified on etc.

That was my intention because it was required in the #271. You can easily enable read-only mode in add form as well by removing `this.$route.params.rowID !== undefined` from the `isReadOnly` method in `RowForm.vue` but I don't think it's necessary because the default value is given by default and you can change it or not, and read-only columns are usually to prevent data editing. Sorry if I missed your point, but feel free to change it, if you think that's better.

**Pull request closed without merging**